bitcoin/src/bip32.rs

```diff
@@ -256,94 +350,237 @@ impl serde::Serialize for ChildNumber {
+    );
+
+    impl NormalDerivationPath {
+        /// Create a new `NormalDerivationPath` from a vector of `NormalChildNumber`
+        pub fn new(vec: Vec<NormalChildNumber>) -> Self { NormalDerivationPath(vec) }
+        /// Returns length of the `NormalDerivationPath`
+        pub fn len(&self) -> usize { self.0.len() }
+        /// Convert `NormalDerivationPath` to a `DerivationPath`
+        pub fn to_derivation_path(&self) -> DerivationPath {
+            self.into_iter().map(|n| n.to_number()).collect()
+        }
 
-#[cfg(feature = "serde")]
-internals::serde_string_impl!(DerivationPath, "a BIP-32 derivation path");
+        /// Returns `true` if the `NormalDerivationPath` is empty
+        pub fn is_empty(&self) -> bool { self.0.is_empty() }
 
-impl<I> Index<I> for DerivationPath
-where
-    Vec<ChildNumber>: Index<I>,
-{
-    type Output = <Vec<ChildNumber> as Index<I>>::Output;
+        /// Returns `NormalDerivationPath` for a master key (i.e. empty `NormalDerivationPath`)
+        pub fn master() -> Self { NormalDerivationPath(vec![]) }
 
-    #[inline]
-    fn index(&self, index: I) -> &Self::Output { &self.0[index] }
-}
+        /// Returns whether `NormalDerivationPath` represents master key (i.e. it's length
+        /// is empty). True for `m` path.
+        pub fn is_master(&self) -> bool { self.0.is_empty() }
 
-impl Default for DerivationPath {
-    fn default() -> DerivationPath { DerivationPath::master() }
-}
+        /// Create a new [NormalDerivationPath] that is a child of this one.
+        pub fn child(&self, cn: NormalChildNumber) -> Self {
+            let mut path = self.0.clone();
+            path.push(cn);
+            NormalDerivationPath(path)
+        }
 
-impl<T> IntoDerivationPath for T
-where
-    T: Into<DerivationPath>,
-{
-    fn into_derivation_path(self) -> Result<DerivationPath, Error> { Ok(self.into()) }
+        /// Convert into a [NormalDerivationPath] that is a child of this one.
+        pub fn into_child(self, cn: NormalChildNumber) -> Self {
+            let mut path = self.0;
+            path.push(cn);
+            NormalDerivationPath(path)
+        }
+        /// Concatenate `self` with `path` and return the resulting new path.
+        ///
+        /// ```
+        /// use bitcoin::bip32::{NormalDerivationPath, NormalChildNumber};
+        /// use std::str::FromStr;
+        ///
+        /// let base = NormalDerivationPath::from_str("m/42").unwrap();
+        ///
+        /// let deriv_1 = base.extend(NormalDerivationPath::from_str("0/1").unwrap());
+        /// let deriv_2 = base.extend(&[
+        ///     NormalChildNumber::ZERO,
+        ///     NormalChildNumber::ONE
+        /// ]);
+        ///
+        /// assert_eq!(deriv_1, deriv_2);
+        /// ```
+        pub fn extend<T: AsRef<[NormalChildNumber]>>(&self, path: T) -> Self {
+            let mut new_path = self.clone();
+            new_path.0.extend_from_slice(path.as_ref());
+            new_path
+        }
```

Comment:
`self.into_iter().chain(path.as_ref()).collect()` doesnt solves the problem ?  `Extend` trait is for the cases that you want to extend a already existent vector. The function wants a new one.

Reply:
I don't see the point of creating a new one, especially since you're not optimizing allocations. Further I think the method should be called `join` like on `Path`.